test_install/test_install.cpp

```diff
@@ -42,13 +44,114 @@ OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+                             const gko::matrix::Dense<> *m2, double prec)
+{
+    assert(m1->get_size()[0] == m2->get_size()[0]);
+    assert(m1->get_size()[1] == m2->get_size()[1]);
+    for (gko::size_type i = 0; i < m1->get_size()[0]; ++i) {
+        for (gko::size_type j = 0; j < m2->get_size()[1]; ++j) {
+            assert(std::abs(m1->at(i, j) - m2->at(i, j)) < prec);
+        }
+    }
+}
+
+
+template <typename Mtx>
+void check_spmv(std::shared_ptr<gko::Executor> exec,
+                gko::matrix_data<double> &A_raw, gko::matrix::Dense<> *b,
+                gko::matrix::Dense<> *x)
+{
+    auto test = Mtx::create(exec);
+#if HAS_REFERENCE
+    auto x_clone = gko::clone(x);
+    test->read(A_raw);
+    test->apply(b, gko::lend(x_clone));
+
+#if defined(HAS_HIP) || defined(HAS_CUDA)
+    auto test_ref = Mtx::create(exec->get_master());
+    auto x_ref = gko::clone(exec->get_master(), x);
+    test_ref->read(A_raw);
+    test_ref->apply(b, gko::lend(x_ref));
+
+    auto x_clone_ref = gko::clone(exec->get_master(), gko::lend(x_clone));
+    assert_similar_matrices(gko::lend(x_clone_ref), gko::lend(x_ref), 1e-14);
+#endif  // defined(HAS_HIP) || defined(HAS_CUDA)
+#endif  // HAS_REFERENCE
+}
+
+
+template <typename Solver>
+void check_solver(std::shared_ptr<gko::Executor> exec,
+                  gko::matrix_data<double> &A_raw, gko::matrix::Dense<> *b,
+                  gko::matrix::Dense<> *x)
+{
+    using Mtx = gko::matrix::Csr<>;
+    auto A =
+        gko::share(Mtx::create(exec, std::make_shared<Mtx::load_balance>()));
+#if HAS_REFERENCE
+    A->read(A_raw);
+#endif  // HAS_REFERENCE
+
+    auto num_iters = 20u;
+    double reduction_factor = 1e-7;
+    auto solver_gen =
+        Solver::build()
+            .with_criteria(
+                gko::stop::Iteration::build().with_max_iters(num_iters).on(
+                    exec),
+                gko::stop::ResidualNorm<>::build()
+                    .with_reduction_factor(reduction_factor)
+                    .on(exec))
+            .on(exec);
+#if HAS_REFERENCE
+    auto x_clone = gko::clone(x);
+    solver_gen->generate(A)->apply(b, gko::lend(x_clone));
+
+#if defined(HAS_HIP) || defined(HAS_CUDA)
+    auto A_ref =
+        gko::share(Mtx::create(exec, std::make_shared<Mtx::load_balance>()));
+    A_ref->read(A_raw);
+    auto refExec = exec->get_master();
+    auto solver_gen_ref =
+        Solver::build()
+            .with_criteria(
+                gko::stop::Iteration::build().with_max_iters(num_iters).on(
+                    exec),
+                gko::stop::ResidualNorm<>::build()
+                    .with_reduction_factor(reduction_factor)
+                    .on(refExec))
+            .on(refExec);
+    auto x_ref = gko::clone(exec->get_master(), x);
+    solver_gen->generate(A_ref)->apply(b, gko::lend(x_ref));
+
+    auto x_clone_ref = gko::clone(exec->get_master(), gko::lend(x_clone));
+    assert_similar_matrices(gko::lend(x_clone_ref), gko::lend(x_ref), 1e-12);
```

Comment:
Does it only test the solver on reference and the copy function on CUDA/HIP?

Reply:
I only compare CUDA/HIP to reference. Reference launches the kernels but does not check if they are correct. It's the same for the solvers.

Reply:
doesn't running the solver on cuda/hip?
The smoke test only involve the copy/allocation on devices side. I wonder whether it is too small on the test.